pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/loader/ForwardIndexHandler.java

```diff
@@ -0,0 +1,269 @@
+        return false;
+      }
+
+      return true;
+    } catch (Exception e) {
+      LOGGER.error("Caught exception while changing compression for column: {}", column, e);
+      return false;
+    }
+  }
+
+  private void rewriteRawForwardIndex(String column, SegmentDirectory.Writer segmentWriter,
+      IndexCreatorProvider indexCreatorProvider)
+      throws Exception {
+    Preconditions.checkState(_segmentMetadata.getVersion() == SegmentVersion.v3);
+
+    ColumnMetadata existingColMetadata = _segmentMetadata.getColumnMetadataFor(column);
+    File indexDir = _segmentMetadata.getIndexDir();
+    String segmentName = _segmentMetadata.getName();
+    File inProgress = new File(indexDir, column + ".fwd.inprogress");
+    File fwdIndexFile = new File(indexDir, column + V1Constants.Indexes.RAW_SV_FORWARD_INDEX_FILE_EXTENSION);
+
+    if (!inProgress.exists()) {
+      // Marker file does not exist, which means last run ended normally.
+      // Create a marker file.
+      FileUtils.touch(inProgress);
+    } else {
+      // Marker file exists, which means last run gets interrupted.
+      // Remove inverted index if exists.
+      // For v1 and v2, it's the actual inverted index. For v3, it's the temporary inverted index.
+      FileUtils.deleteQuietly(fwdIndexFile);
+    }
+
+    LOGGER.info("Creating new forward index for segment={} and column={}", segmentName, column);
+
+    Map<String, ChunkCompressionType> compressionConfigs = _indexLoadingConfig.getCompressionConfigs();
+    Preconditions.checkState(compressionConfigs.containsKey(column));
+    // At this point, compressionConfigs is guaranteed to contain the column.
+    ChunkCompressionType newCompressionType = compressionConfigs.get(column);
+
+    int numDocs = existingColMetadata.getTotalDocs();
+
+    try (ForwardIndexReader reader = LoaderUtils.getForwardIndexReader(segmentWriter, existingColMetadata)) {
+      int lengthOfLongestEntry = reader.getLengthOfLongestEntry();
+      Preconditions.checkState(lengthOfLongestEntry >= 0,
+          "lengthOfLongestEntry cannot be negative. segment=" + segmentName + " column={}" + column);
+
+      IndexCreationContext.Forward context =
+          IndexCreationContext.builder().withIndexDir(indexDir).withColumnMetadata(existingColMetadata)
+              .withLengthOfLongestEntry(lengthOfLongestEntry).build()
+              .forForwardIndex(newCompressionType, _indexLoadingConfig.getColumnProperties());
+
+      try (ForwardIndexCreator creator = indexCreatorProvider.newForwardIndexCreator(context)) {
+        PinotSegmentColumnReader columnReader =
+            new PinotSegmentColumnReader(reader, null, null, existingColMetadata.getMaxNumberOfMultiValues());
+
+        for (int i = 0; i < numDocs; i++) {
+          Object val = columnReader.getValue(i);
+
+          // JSON fields are either stored as string or bytes. No special handling is needed.
```

Comment:
Just curious, in the `SegmentColumnarIndexCreator` I see the following for JSON handling:

```
            case JSON:
              if (columnValueToIndex instanceof String) {
                forwardIndexCreator.putString((String) columnValueToIndex);
              } else if (columnValueToIndex instanceof byte[]) {
                forwardIndexCreator.putBytes((byte[]) columnValueToIndex);
              }
              break;
```

Do we not need to do the same here? 

From checking the code it looks like the forward index creator is correctly setup with the stored type: `context.getFieldSpec().getDataType().getStoredType()` so maybe this is indeed not a concern here. Just asking to double check this part.

Not needed for this PR but we should perhaps cleanup this JSON handling part in `SegmentColumnarIndexCreator` as well to avoid confusion.

Reply:
It's not needed here specifically because - as you said we use the storedType. To avoid confusion, I've added a comment for clarification and also used the storedType from reader. This is guaranteed to be either STRING or BYTES for JSON and also easier to understand. 

Not sure why we have the code in `SegmentColumnarIndexCreator`. As this code also creates a forwardIndex context similar to how IndexHandler does, we should never get JSON type. But it was added recently here https://github.com/apache/pinot/pull/8073. @siddharthteotia  any inputs here?